java/flight/flight-sql-jdbc-core/src/main/java/org/apache/arrow/driver/jdbc/ArrowFlightJdbcFlightStreamResultSet.java

```diff
@@ -130,39 +146,89 @@ protected AvaticaResultSet execute() throws SQLException {
   }
 
   private void execute(final FlightInfo flightInfo) throws SQLException {
-    loadNewQueue();
+    // load new FlightStreamQueue
+    ofNullable(flightStreamQueue).ifPresent(AutoCloseables::closeNoChecked);
+    flightStreamQueue = createNewQueue(connection.getExecutorService());
+
+    // load new FlightStream
     flightStreamQueue.enqueue(connection.getClientHandler().getStreams(flightInfo));
-    loadNewFlightStream();
+    ofNullable(currentFlightStream).ifPresent(AutoCloseables::closeNoChecked);
+    currentFlightStream = getNextFlightStream(true);
 
     // Ownership of the root will be passed onto the cursor.
     if (currentFlightStream != null) {
-      executeForCurrentFlightStream();
+      storeRootsFromStreamAsync();
+      executeNextRoot();
+    }
+  }
+
+  private BufferAllocator getAllocator() {
+    if (allocator == null) {
+      allocator = connection.getBufferAllocator().newChildAllocator("vsr-copier", 0, Long.MAX_VALUE);
     }
+
+    return allocator;
   }
 
-  private void executeForCurrentFlightStream() throws SQLException {
-    final VectorSchemaRoot originalRoot = currentFlightStream.getRoot();
+  private VectorSchemaRoot cloneRoot(VectorSchemaRoot originalRoot) {
+    VectorSchemaRoot theRoot = VectorSchemaRoot.create(originalRoot.getSchema(), getAllocator());
+    VectorLoader loader = new VectorLoader(theRoot);
+    VectorUnloader unloader = new VectorUnloader(originalRoot);
+    try (ArrowRecordBatch recordBatch = unloader.getRecordBatch()) {
+      loader.load(recordBatch);
+    }
+    return theRoot;
+  }
 
+  private void storeRoot(VectorSchemaRoot originalRoot) throws SQLException {
+    VectorSchemaRoot theRoot = cloneRoot(originalRoot);
+    VectorSchemaRoot transformedRoot = null;
     if (transformer != null) {
       try {
-        currentVectorSchemaRoot = transformer.transform(originalRoot, currentVectorSchemaRoot);
+        transformedRoot = transformer.transform(theRoot, null);
+        theRoot.close();
       } catch (final Exception e) {
         throw new SQLException("Failed to transform VectorSchemaRoot.", e);
       }
-    } else {
-      currentVectorSchemaRoot = originalRoot;
     }
 
-    if (schema != null) {
-      execute(currentVectorSchemaRoot, schema);
-    } else {
-      execute(currentVectorSchemaRoot);
+    try {
+      vectorSchemaRoots.put(ofNullable(transformedRoot).orElse(theRoot));
+    } catch (InterruptedException e) {
+      throw new SQLException("Could not put root to the queue", e);
     }
   }
 
+  private void executeNextRoot() throws SQLException {
+    try {
+      ofNullable(currentRoot).ifPresent(AutoCloseables::closeNoChecked);
+      currentRoot = vectorSchemaRoots.poll(10, TimeUnit.SECONDS);
```

Comment:
Do we want to have a hard timeout like this? We don't know how long it might take for the server to respond